### PR TITLE
LTR-303, LTR-329, LTR-553, LTR-556, LTR-559, LTR-659 Series of Lite-On Light (ALS) and Proximity(PS) sensors

### LINKED ISSUE #2485: Support for sensor: LTR-303
**Describe the problem you have/What new integration you would like**
I would like the Adafruit Light Sensor "LTR-303" to be integrated to ESPHome. I suspect it would be similar to the integration that already exists named "LTR-390"

**Please describe your use case for this integration and alternatives you've tried:**
1. The integration does not exist with ESPHome
2. I've tried using Micropython, but there are no libraries for it.
3. I can't use CircuitPython, since ESP8266 is no longer supported

**Additional context**
N/A

### LINKED ISSUE #2551: Add support for Lite-On LTR-553 illuminance & proximity sensor for CoreS3
**Add support for Lite-On LTR-553 proximity sensor for CoreS3**

**Please describe your use case for this integration and alternatives you've tried:**

[CoreS3](https://docs.m5stack.com/en/core/CoreS3) is the third generation of the M5Stack Core series. Camera GC0308 is integrated at the bottom of the screen, accompanied by a proximity sensor [LTR-553ALS-WA ](https://optoelectronics.liteon.com/upload/download/DS86-2014-0007/LTR-553ALS-01_DS_V1.pdf)for enhanced functionality.

 Ardunio lib: [LTR507-LTR553-LTR556](https://github.com/chenhonglin/LTR507-LTR553-LTR556_Arduino/blob/master/README.md)

**Additional context**

Datasheet: [LTR-553ALS-WA](https://m5stack.oss-cn-shenzhen.aliyuncs.com/resource/docs/datasheet/core/K128%20CoreS3/LTR-553ALS-WA.PDF)

image

### PR DESCRIPTION
**Pull request by latonita**
# What does this implement/fix?

Component for using LiteOn LTR-303, LTR-329, LTR-553, LTR-556, LTR-559, LTR-659 ambient light and proximity sensors.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/2551
and also fixes https://github.com/esphome/feature-requests/issues/2485

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3528

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [X] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
esphome:
  name: ltr303-test

esp32:
  board: esp32dev
  framework:
    type: arduino

# esp8266:
#   board: nodemcuv2

logger:
  level: DEBUG

i2c:
#  sda: GPIO25
#  scl: GPIO32
#  scan: true

sensor:
  - platform: ltr_als_ps
    address: 0x23
    type: ALS_PS
    auto_mode: true
    gain: 1x
    integration_time: 100ms
    ps_cooldown: 5 s
    ambient_light:
      name: "Ambient light"
    ps_counts:
      name: "Proximity counts"

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/ltr303/ltr303.cpp [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+    if (current_gain != gains[gain_idx]) {
+      this->configure_gain_(gains[gain_idx]);
+      current_gain = gains[gain_idx];
+      wait_needed = true;
+    }
+    // wait for new data samples to come
+    // delay(ALS_INT_TIME[current_time] * 2);
+    if (wait_needed) {
+      delay(ALS_MEAS_RATE[this->repeat_rate_]);
```
Comment (kbx81) [inline esphome/components/ltr303/ltr303.cpp]: As Jesse mentioned [here](https://github.com/esphome/esphome/pull/6067#discussion_r1457026808), we can't have big delays like this. Please rework to eliminate this.

File: esphome/components/ltr303/ltr303.cpp [hunk 2]
```diff
@@ -0,0 +1,322 @@
+#include "ltr303.h"
+#include "esphome/core/application.h"
+#include "esphome/core/log.h"
+#include "esphome/core/helpers.h"
+
+namespace esphome {
+namespace ltr303 {
+
+static const char *const TAG = "ltr303";
+static const int ALS_GAIN[8] = {1, 2, 4, 8, 0, 0, 48, 96};                        // multiplier
```
Comment (kbx81) [inline esphome/components/ltr303/ltr303.cpp]: These could be `uint8_t` -- would save some memory...
```suggestion
static const uint8_t ALS_GAIN[8] = {1, 2, 4, 8, 0, 0, 48, 96};                        // multiplier
```

File: esphome/components/ltr303/ltr303.cpp [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+static const int ALS_INT_TIME[8] = {100, 50, 200, 400, 150, 250, 300, 350};       // ms
+static const int ALS_MEAS_RATE[8] = {50, 100, 200, 500, 1000, 2000, 2000, 2000};  // ms
```
Comment (kbx81) [inline esphome/components/ltr303/ltr303.cpp]: As above, these could save some memory by making them `uint16_t`.
```suggestion
static const uint16_t ALS_INT_TIME[8] = {100, 50, 200, 400, 150, 250, 300, 350};       // ms
static const uint16_t ALS_MEAS_RATE[8] = {50, 100, 200, 500, 1000, 2000, 2000, 2000};  // ms
```

File: esphome/components/ltr303/ltr303.h [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  CR_ALS_CTRL = 0x80,   // ALS operation mode control SW reset
+  CR_MEAS_RATE = 0x85,  // ALS measurement rate in active mode
+  CR_PART_ID = 0x86,    // Part Number ID and Revision ID
+  CR_MANU_ID = 0x87,    // Manufacturer ID
+  CR_CH1_0 = 0x88,      // ALS measurement CH1 data, lower byte - infrared only
+  CR_CH1_1 = 0x89,      // ALS measurement CH1 data, upper byte - infrared only
+  CR_CH0_0 = 0x8A,      // ALS measurement CH0 data, lower byte - visible + infrared
+  CR_CH0_1 = 0x8B,      // ALS measurement CH0 data, upper byte - visible + infrared
+  CR_ALS_STATUS = 0x8c  // ALS new data status
+};
```
Comment (kbx81) [inline esphome/components/ltr303/ltr303.h]: In ESPHome, we typically use (static) consts for this purpose.
```suggestion
static const uint8_t CR_ALS_CTRL = 0x80;   // ALS operation mode control SW reset
static const uint8_t CR_MEAS_RATE = 0x85;  // ALS measurement rate in active mode
static const uint8_t CR_PART_ID = 0x86;    // Part Number ID and Revision ID
static const uint8_t CR_MANU_ID = 0x87;    // Manufacturer ID
static const uint8_t CR_CH1_0 = 0x88;      // ALS measurement CH1 data, lower byte - infrared only
static const uint8_t CR_CH1_1 = 0x89;      // ALS measurement CH1 data, upper byte - infrared only
static const uint8_t CR_CH0_0 = 0x8A;      // ALS measurement CH0 data, lower byte - visible + infrared
static const uint8_t CR_CH0_1 = 0x8B;      // ALS measurement CH0 data, upper byte - visible + infrared
static const uint8_t CR_ALS_STATUS = 0x8c; // ALS new data status
```

**Review by kbx81 — CHANGES_REQUESTED** [review]
Thank you for working on this -- just a few comments.

File: esphome/components/ltr303/ltr303.h (hunk 4 above)
Comment (latonita) [reply esphome/components/ltr303/ltr303.h]: yes, i saw this usage in other components, but I think we might move forward small steps at least.
Simple `static const` are really from the old C-style past. 
I believe enums give us much more clearer and readable code, especially if used prefixed with enum name like `RegisterMap::CONF_ABC`. Also it works as typecheck.
If there is one or two values - consts are okay, 
when we have numbers of them - semantically its better to organize them into enums.

I'm strong believer we shall keep with enums, even if it (yet) looks a little bit different than old approach. All in all its not of that big difference in the end, but readability much better.
What do you think? 

I have few more components I'm in progress of making now. All with enums now, really easier to read.  (VEML7700, INA228).

File: esphome/components/ltr303/ltr303.cpp (hunk 1 above)
Comment (latonita) [reply esphome/components/ltr303/ltr303.cpp]: I will rework this one a little bit similarly to VEML7700 (state machine in the loop)

File: esphome/components/ltr303/ltr303.h (hunk 4 above)
Comment (kbx81) [reply esphome/components/ltr303/ltr303.h]: I'm not strictly opposed to it if it's a generally-accepted new(ish) approach -- would be curious if @jesserockz has an opinion. I've seen it come up elsewhere a couple of times (outside of your PRs) in the last few weeks but didn't realize it's a new pattern people are adopting.

File: esphome/components/ltr303/ltr303.cpp (hunk 2 above)
Comment (latonita) [reply esphome/components/ltr303/ltr303.cpp]: thats my bad, copy paste from my test code. fixed

**Comment by latonita** [comment]
Looks like this component might support more than 2 sensors. There is number of ALS sensors, PS(Proximity) sensors and ALS + PS integrated sensors from Lite-On which share same API, the only issue they can't be detected by product_id, its just the same form many of them with different features, so manual selection will be required.
I'm reworking component, it will be `ltr_alsps` supporting at least LTR-303,329, 553, 556, 659 and some others

**Comment by Donkie** [comment]
I have been trying this for some time with very mixed results, none of them I would call successful.
I previously used the exact same board and sensor etc but with a simple arduino code to talk to the sensor. Then it worked fine and I got proper readings.

However with this esphome sensor I either get extremely noisy data or incorrect data. I tried it first using the auto setting, and then it was very noisy during daylight, it kept jumping up and down from 0 to 1000s of lux:
image

Now with auto off, I some days get very low values like this:
image

Here is a log from today, where the values was slightly higher, but still much lower than they were with my custom code.
```
INFO ESPHome 2023.12.9
INFO Reading configuration /config/weather-station.yaml...
INFO Updating https://github.com/latonita/esphome-ltr303.git@None
INFO Detected timezone 'Europe/Berlin'
[10:25:13][I][app:102]: ESPHome version 2023.12.9 compiled on Jan 26 2024, 17:02:16
[10:25:13][C][status_led:019]: Status LED:
[10:25:13][C][status_led:020]:   Pin: GPIO16
[10:25:13][C][wifi:573]: WiFi:
[10:25:13][C][wifi:405]:  
[10:25:13][C][logger:439]: Logger:
[10:25:13][C][logger:440]:   Level: DEBUG
[10:25:13][C][logger:441]:   Log Baud Rate: 115200
[10:25:13][C][logger:443]:   Hardware UART: UART0
[10:25:13][C][i2c.arduino:053]: I2C Bus:
[10:25:13][C][i2c.arduino:054]:   SDA Pin: GPIO4
[10:25:13][C][i2c.arduino:055]:   SCL Pin: GPIO5
[10:25:13][C][i2c.arduino:056]:   Frequency: 50000 Hz
[10:25:13][C][i2c.arduino:059]:   Recovery: bus successfully recovered
[10:25:13][I][i2c.arduino:069]: Results from i2c bus scan:
[10:25:13][I][i2c.arduino:075]: Found i2c device at address 0x29
[10:25:13][C][uptime.sensor:031]: Uptime Sensor 'Uptime'
[10:25:13][C][uptime.sensor:031]:   Device Class: 'duration'
[10:25:13][C][uptime.sensor:031]:   State Class: 'total_increasing'
[10:25:13][C][uptime.sensor:031]:   Unit of Measurement: 's'
[10:25:13][C][uptime.sensor:031]:   Accuracy Decimals: 0
[10:25:13][C][uptime.sensor:031]:   Icon: 'mdi:timer-outline'
[10:25:13][C][homeassistant.time:010]: Home Assistant Time:
[10:25:13][C][homeassistant.time:011]:   Timezone: 'CET-1CEST,M3.5.0,M10.5.0/3'
[10:25:13][C][ltr303:067]:   Address: 0x29
[10:25:13][C][ltr303:068]:   Automatic mode: OFF
[10:25:13][C][ltr303:069]:   Gain: 1x
[10:25:13][C][ltr303:070]:   Integration time: 400 ms
[10:25:13][C][ltr303:071]:   Measurement repeat rate: 500 ms
[10:25:13][C][ltr303:072]:   Glass attenuation factor: 1.000000
[10:25:13][C][ltr303:073]:   Update Interval: 60.0s
[10:25:13][C][ltr303:075]:   ALS calculated lux 'Ambient light'
[10:25:13][C][ltr303:075]:     Device Class: 'illuminance'
[10:25:13][C][ltr303:075]:     State Class: 'measurement'
[10:25:13][C][ltr303:075]:     Unit of Measurement: 'lx'
[10:25:13][C][ltr303:075]:     Accuracy Decimals: 1
[10:25:13][C][ltr303:075]:     Icon: 'mdi:brightness-6'
[10:25:13][C][ltr303:076]:   CH1 Infrared counts 'Infrared counts'
[10:25:13][C][ltr303:076]:     Device Class: 'illuminance'
[10:25:13][C][ltr303:076]:     State Class: 'measurement'
[10:25:13][C][ltr303:076]:     Unit of Measurement: '#'
[10:25:13][C][ltr303:076]:     Accuracy Decimals: 0
[10:25:13][C][ltr303:076]:     Icon: 'mdi:brightness-5'
[10:25:13][C][ltr303:077]:   CH0 Visible+IR counts 'Full spectrum counts'
[10:25:13][C][ltr303:077]:     Device Class: 'illuminance'
[10:25:13][C][ltr303:077]:     State Class: 'measurement'
[10:25:13][C][ltr303:077]:     Unit of Measurement: '#'
[10:25:13][C][ltr303:077]:     Accuracy Decimals: 0
[10:25:13][C][ltr303:077]:     Icon: 'mdi:brightness-7'
[10:25:13][C][ltr303:078]:   Actual gain 'Actual gain'
[10:25:13][C][ltr303:078]:     Device Class: 'illuminance'
[10:25:13][C][ltr303:078]:     State Class: 'measurement'
[10:25:13][C][ltr303:078]:     Unit of Measurement: ''
[10:25:13][C][ltr303:078]:     Accuracy Decimals: 0
[10:25:13][C][ltr303:078]:     Icon: 'mdi:multiplication'
[10:25:13][C][dht:017]: DHT:
[10:25:13][C][dht:018]:   Pin: GPIO12
[10:25:13][C][dht:020]:   Auto-detected model: DHT22
[10:25:13][C][dht:027]:   Update Interval: 60.0s
[10:25:13][C][dht:029]:   Temperature 'Temperature'
[10:25:13][C][dht:029]:     Device Class: 'temperature'
[10:25:13][C][dht:029]:     State Class: 'measurement'
[10:25:13][C][dht:029]:     Unit of Measurement: '°C'
[10:25:13][C][dht:029]:     Accuracy Decimals: 1
[10:25:13][C][dht:030]:   Humidity 'Humidity'
[10:25:13][C][dht:030]:     Device Class: 'humidity'
[10:25:13][C][dht:030]:     State Class: 'measurement'
[10:25:13][C][dht:030]:     Unit of Measurement: '%'
[10:25:13][C][dht:030]:     Accuracy Decimals: 0
[10:25:13][C][captive_portal:088]: Captive Portal:
[10:25:13][C][mdns:115]: mDNS:
[10:25:13][C][mdns:116]:   Hostname: weather-station
[10:25:13][C][ota:097]: Over-The-Air Updates:
[10:25:13][C][ota:098]:   Address: weather-station.local:8266
[10:25:13][C][ota:101]:   Using Password.
[10:25:13][C][api:139]: API Server:
[10:25:13][C][api:140]:   Address: weather-station.local:6053
[10:25:13][C][api:142]:   Using noise encryption: YES
[10:25:13][C][wifi_signal.sensor:009]: WiFi Signal 'RSSI'
[10:25:13][C][wifi_signal.sensor:009]:   Device Class: 'signal_strength'
[10:25:13][C][wifi_signal.sensor:009]:   State Class: 'measurement'
[10:25:13][C][wifi_signal.sensor:009]:   Unit of Measurement: 'dBm'
[10:25:13][C][wifi_signal.sensor:009]:   Accuracy Decimals: 0
[10:25:27][D][ltr303:086]: Updating
[10:25:27][D][ltr303:088]: Initiating new data collection
[10:25:27][D][ltr303:244]: Data ready, reported gain is 1
[10:25:27][D][ltr303:126]: Reading sensor data having gain = 1x, time = 400 ms
[10:25:27][D][ltr303:256]: Got sensor data: CH1 = 150, CH0 = 89
[10:25:27][D][ltr303:328]: Lux calculation: ratio 0.627615, gain 1.000000, int time 4.000000, inv_pfactor 1.000000
[10:25:27][D][sensor:093]: 'Ambient light': Sending state 21.89162 lx with 1 decimals of accuracy
[10:25:27][D][sensor:093]: 'Infrared counts': Sending state 150.00000 # with 0 decimals of accuracy
[10:25:27][D][sensor:093]: 'Full spectrum counts': Sending state 89.00000 # with 0 decimals of accuracy
[10:25:27][D][sensor:093]: 'Actual gain': Sending state 1.00000  with 0 decimals of accuracy
[10:25:27][D][sensor:093]: 'Actual integration time': Sending state 400.00000 ms with 0 decimals of accuracy
[10:25:31][D][dht:048]: Got Temperature=4.7°C Humidity=99.9%
[10:25:32][D][sensor:093]: 'Temperature': Sending state 4.70000 °C with 1 decimals of accuracy
[10:25:32][D][sensor:093]: 'Humidity': Sending state 99.90000 % with 0 decimals of accuracy
[10:25:36][D][sensor:093]: 'Uptime': Sending state 148969.73438 s with 0 decimals of accuracy
[10:25:53][D][sensor:093]: 'RSSI': Sending state -65.00000 dBm with 0 decimals of accuracy
[10:26:27][D][ltr303:086]: Updating
[10:26:27][D][ltr303:088]: Initiating new data collection
[10:26:27][D][ltr303:244]: Data ready, reported gain is 1
[10:26:27][D][ltr303:126]: Reading sensor data having gain = 1x, time = 400 ms
[10:26:27][D][ltr303:256]: Got sensor data: CH1 = 125, CH0 = 76
[10:26:27][D][ltr303:328]: Lux calculation: ratio 0.621891, gain 1.000000, int time 4.000000, inv_pfactor 1.000000
[10:26:27][D][sensor:093]: 'Ambient light': Sending state 20.20400 lx with 1 decimals of accuracy
[10:26:27][D][sensor:093]: 'Infrared counts': Sending state 125.00000 # with 0 decimals of accuracy
[10:26:27][D][sensor:093]: 'Full spectrum counts': Sending state 76.00000 # with 0 decimals of accuracy
[10:26:27][D][sensor:093]: 'Actual gain': Sending state 1.00000  with 0 decimals of accuracy
[10:26:27][D][sensor:093]: 'Actual integration time': Sending state 400.00000 ms with 0 decimals of accuracy
[10:26:31][D][dht:048]: Got Temperature=4.7°C Humidity=99.9%
[10:26:32][D][sensor:093]: 'Temperature': Sending state 4.70000 °C with 1 decimals of accuracy
[10:26:32][D][sensor:093]: 'Humidity': Sending state 99.90000 % with 0 decimals of accuracy
[10:26:36][D][sensor:093]: 'Uptime': Sending state 149029.73438 s with 0 decimals of accuracy
[10:26:53][D][sensor:093]: 'RSSI': Sending state -65.00000 dBm with 0 decimals of accuracy
[10:27:10][D][time:044]: Synchronized time: 2024-01-28 10:27:10
[10:27:11][D][time:044]: Synchronized time: 2024-01-28 10:27:10
[10:27:27][D][ltr303:086]: Updating
[10:27:27][D][ltr303:088]: Initiating new data collection
[10:27:27][D][ltr303:244]: Data ready, reported gain is 1
[10:27:27][D][ltr303:126]: Reading sensor data having gain = 1x, time = 400 ms
[10:27:27][D][ltr303:256]: Got sensor data: CH1 = 107, CH0 = 67
[10:27:27][D][ltr303:328]: Lux calculation: ratio 0.614943, gain 1.000000, int time 4.000000, inv_pfactor 1.000000
[10:27:27][D][sensor:093]: 'Ambient light': Sending state 19.37398 lx with 1 decimals of accuracy
[10:27:27][D][sensor:093]: 'Infrared counts': Sending state 107.00000 # with 0 decimals of accuracy
[10:27:27][D][sensor:093]: 'Full spectrum counts': Sending state 67.00000 # with 0 decimals of accuracy
[10:27:27][D][sensor:093]: 'Actual gain': Sending state 1.00000  with 0 decimals of accuracy
[10:27:27][D][sensor:093]: 'Actual integration time': Sending state 400.00000 ms with 0 decimals of accuracy
[10:27:32][D][dht:048]: Got Temperature=4.7°C Humidity=99.9%
[10:27:32][D][sensor:093]: 'Temperature': Sending state 4.70000 °C with 1 decimals of accuracy
[10:27:32][D][sensor:093]: 'Humidity': Sending state 99.90000 % with 0 decimals of accuracy
[10:27:36][D][sensor:093]: 'Uptime': Sending state 149089.71875 s with 0 decimals of accuracy
```

It might be that this is an issue from my side, I'm not fully sure yet. I can do some more testing and comparing my code with this if necessary, but I just wanted to call out if anyone else experienced similar issues.

LTR-303 with a NodeMCU host

**Comment by latonita** [comment]
> It might be that this is an issue from my side, I'm not fully sure yet. I can do some more testing and comparing my code with this if necessary, but I just wanted to call out if anyone else experienced similar issues.
> 
> LTR-303 with a NodeMCU host

Hi. Thanks for reporting, would you please post follow up in this thread https://github.com/esphome/feature-requests/issues/2485 
What I see is that data coming from sensor looks ok. If you can compare manual results of this component and some arduino library you using - it would be great. here https://github.com/esphome/feature-requests/issues/2485 
thanks a lot!

File: esphome/components/ltr303/ltr303.h (hunk 4 above)
Comment (latonita) [reply esphome/components/ltr303/ltr303.h]: resolving as 'happy to disagree', moreover a lot of other things changed as well

**Comment by Donkie** [comment]
> > It might be that this is an issue from my side, I'm not fully sure yet. I can do some more testing and comparing my code with this if necessary, but I just wanted to call out if anyone else experienced similar issues.
> > LTR-303 with a NodeMCU host
> 
> Hi. Thanks for reporting, would you please post follow up in this thread [esphome/feature-requests#2485](https://github.com/esphome/feature-requests/issues/2485) What I see is that data coming from sensor looks ok. If you can compare manual results of this component and some arduino library you using - it would be great. here [esphome/feature-requests#2485](https://github.com/esphome/feature-requests/issues/2485) thanks a lot!

My issues seems to have mysteriously fixed themselves, so I don't think it's an issue with the implementation, mostly an issue on my side. Thanks anyway :)

File: tests/test2.yaml [hunk 5]
```diff
@@ -311,6 +311,20 @@ sensor:
...
+    ps_cooldown: 5 s
+    ambient_light:
+      name: "Ambient light"
+    full_spectrum_counts:
+      name: "Full spectrum counts"
+    infrared_counts:
+      name: "Infrared counts"
+    actual_gain:
+      name: "Actual gain"
+    update_interval: 60s
```
Comment (kbx81) [inline tests/test2.yaml]: Not needed with the new tests in place
```suggestion
```

Comment (latonita) [reply tests/test2.yaml]: old-style test removed

File: esphome/components/ltr_als_ps/ltr_als_ps.cpp [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+  return ALS_GAIN[gain & 0b111];
+}
+
+static float get_ps_gain_coeff(PsGain gain) {
+  static const float PS_GAIN[4] = {16, 0, 32, 64};
+  return PS_GAIN[gain & 0b11];
+}
+
+void LTRAlsPsComponent::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up LTR-303/329");
```
Comment (kbx81) [inline esphome/components/ltr_als_ps/ltr_als_ps.cpp]: I wonder if this should be expanded somehow to included all supported sensors; maybe something like this?:
```suggestion
  ESP_LOGCONFIG(TAG, "Setting up LTR-303/329/55x/659");
```

File: esphome/components/ltr_als_ps/ltr_als_ps.cpp [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+
+void LTRAlsPsComponent::dump_config() {
+  auto get_device_type = [](LtrType typ) {
+    switch (typ) {
+      case LtrType::LTR_TYPE_ALS_ONLY:
+        return "ALS only";
+      case LtrType::LTR_TYPE_PS_ONLY:
+        return "PS only";
+      case LtrType::LTR_TYPE_ALS_AND_PS:
+        return "Als + PS";
```
Comment (kbx81) [inline esphome/components/ltr_als_ps/ltr_als_ps.cpp]: Make this consistent with what's used above
```suggestion
        return "ALS + PS";
```

File: esphome/components/ltr_als_ps/ltr_als_ps.cpp [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+    ESP_LOGCONFIG(TAG, "  Proximity gain: %.0fx", get_ps_gain_coeff(this->ps_gain_));
+    ESP_LOGCONFIG(TAG, "  Proximity cooldown time: %d s", this->ps_cooldown_time_s_);
+    ESP_LOGCONFIG(TAG, "  Proximity high threshold: %d", this->ps_threshold_high_);
+    ESP_LOGCONFIG(TAG, "  Proximity low threshold: %d", this->ps_threshold_low_);
+    LOG_SENSOR("  ", "Proximity counts", this->proximity_counts_sensor_);
+  }
+  LOG_UPDATE_INTERVAL(this);
+
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Communication with I2C LTR-303/329 failed!");
```
Comment (kbx81) [inline esphome/components/ltr_als_ps/ltr_als_ps.cpp]: If we change what's logged in `setup()`...just a callout to keep this consistent.
```suggestion
    ESP_LOGE(TAG, "Communication with I2C LTR-303/329/55x/659 failed!");
```

File: esphome/components/ltr_als_ps/ltr_als_ps.cpp [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+  ESP_LOGV(TAG, "Resetting");
+
+  AlsControlRegister als_ctrl{0};
+  als_ctrl.sw_reset = true;
+  this->reg((uint8_t) CommandRegisters::ALS_CONTR) = als_ctrl.raw;
+  delay(2);
+
+  uint8_t tries = MAX_TRIES;
+  do {
+    ESP_LOGV(TAG, "Waiting chip to reset");
```
Comment (kbx81) [inline esphome/components/ltr_als_ps/ltr_als_ps.cpp]:
```suggestion
    ESP_LOGV(TAG, "Waiting for chip to reset");
```

File: esphome/components/ltr_als_ps/ltr_als_ps.cpp [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+    delay(2);
+    als_ctrl.raw = this->reg((uint8_t) CommandRegisters::ALS_CONTR).get();
+  } while (als_ctrl.sw_reset && tries--);  // while sw reset bit is on - keep waiting
+
+  if (als_ctrl.sw_reset) {
+    ESP_LOGW(TAG, "Failed to finalize reset procedure");
```
Comment (kbx81) [inline esphome/components/ltr_als_ps/ltr_als_ps.cpp]: I think I would just say:
```suggestion
    ESP_LOGW(TAG, "Reset timed out");
```

File: esphome/components/ltr_als_ps/ltr_als_ps.cpp [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+  if (!als_status.ps_new_data || als_status.data_invalid) {
+    return this->ps_readings_;
+  }
+
+  uint8_t ps_low = this->reg((uint8_t) CommandRegisters::PS_DATA_0).get();
+  PsData1Register ps_high;
+  ps_high.raw = this->reg((uint8_t) CommandRegisters::PS_DATA_1).get();
+
+  uint16_t val = encode_uint16(ps_high.ps_data_high, ps_low);
+  //  ESP_LOGV(TAG, "Got sensor data: PS = %5d, Saturation flag = %d", val, ps_high.ps_saturation_flag);
```
Comment (kbx81) [inline esphome/components/ltr_als_ps/ltr_als_ps.cpp]: If you don't want this, please delete -- alternatively, you could make it an `ESP_LOGVV` (very verbose).

File: esphome/components/ltr_als_ps/ltr_als_ps.cpp [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+      return true;
+    }
+    IntegrationTime prev_time = get_prev(INT_TIMES, data.integration_time);
+    if (prev_time != data.integration_time) {
+      data.integration_time = prev_time;
+      ESP_LOGV(TAG, "High illuminance. Decreasing integration time.");
+      return true;
+    }
+  } else {
+    ESP_LOGD(TAG, "Illuminance is good enough.");
```
Comment (kbx81) [inline esphome/components/ltr_als_ps/ltr_als_ps.cpp]: Total nitpick, but I think this looks better. haha
```suggestion
    ESP_LOGD(TAG, "Illuminance is sufficient.");
```

File: esphome/components/ltr_als_ps/ltr_als_ps.cpp [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+  }
+}
+
+void LTRAlsPsComponent::publish_data_part_2_(AlsReadings &data) {
+  if (this->actual_gain_sensor_ != nullptr) {
+    this->actual_gain_sensor_->publish_state(get_gain_coeff(data.gain));
+  }
+  if (this->actual_integration_time_sensor_ != nullptr) {
+    this->actual_integration_time_sensor_->publish_state(get_itime_ms(data.integration_time));
+  }
```
Comment (kbx81) [inline esphome/components/ltr_als_ps/ltr_als_ps.cpp:516]: What is the reason for splitting the state publishing into two parts?

**Review by kbx81 — COMMENTED** [review]
Just a couple little nitpicks/corrections for some messages and a question, but overall I think this is looking good! 😄

File: esphome/components/ltr_als_ps/ltr_als_ps.cpp (hunk 13 above)
Comment (latonita) [reply esphome/components/ltr_als_ps/ltr_als_ps.cpp:516]: With many sensors enabled it takes time to publish and triggers 'component ... took a long time'.
I believe its not a big deal to publish in two batches, just making sure we don't hold the line.

File: esphome/components/ltr_als_ps/ltr_als_ps.cpp (hunk 11 above)
Comment (latonita) [reply esphome/components/ltr_als_ps/ltr_als_ps.cpp]: forgot to remove. removed now.

File: esphome/components/ltr_als_ps/ltr_als_ps.cpp (hunk 13 above)
Comment (kbx81) [reply esphome/components/ltr_als_ps/ltr_als_ps.cpp:516]: I kind of figured -- thanks for confirming. 😄

**Review by kbx81 — APPROVED** [review]

**Pull request merged**